pkg/resolver/search.go

```diff
@@ -186,39 +191,50 @@ func (s *SearchResult) resolveItems() ([]map[string]interface{}, error) {
+	var whereDs []exp.Expression
 
 	if len(s.input.RelatedKinds) > 0 {
-		// TODO: Use the RelatedKinds filter in the SQL query.
-		klog.Warning("TODO: The relationships query must use the provided kind filters.")
+		relatedKinds := pointerToStringArray(s.input.RelatedKinds)
+		whereDs = append(whereDs, goqu.C("destkind").In(relatedKinds).Expression())
 	}
+	//The level can be parameterized later, if needed, for applications
+	whereDs = append(whereDs, goqu.C("level").Eq(1)) // Add filter to select only level 1 relationships
 
 	//defining variables
 	items := []map[string]interface{}{}
 	var kindSlice []string
 	var kindList []string
 	var countList []int
 
-	relQuery := strings.TrimSpace(`WITH RECURSIVE
-	search_graph(uid, data, destkind, sourceid, destid, path, level)
-	AS (
-	SELECT r.uid, r.data, e.destkind, e.sourceid, e.destid, ARRAY[r.uid] AS path, 1 AS level
-		FROM search.resources r
-		INNER JOIN
-			search.edges e ON (r.uid = e.sourceid) OR (r.uid = e.destid)
-		 WHERE r.uid = ANY($1)
-	UNION
-	SELECT r.uid, r.data, e.destkind, e.sourceid, e.destid, path||r.uid, level+1 AS level
-		FROM search.resources r
-		INNER JOIN
-			search.edges e ON (r.uid = e.sourceid)
-		, search_graph sg
-		WHERE (e.sourceid = sg.destid OR e.destid = sg.sourceid)
-		AND r.uid <> all(sg.path)
-		AND level = 1
-		)
-	SELECT distinct ON (destid) data, destid, destkind FROM search_graph WHERE level=1 OR destid = ANY($1)`)
-
-	relations, relQueryError := s.pool.Query(context.Background(), relQuery, s.uids) // how to deal with defaults.
+	schema := goqu.S("search")
+	selectBase := make([]interface{}, 0)
+	selectBase = append(selectBase, "r.uid", "r.data", "e.destkind", "e.sourceid", "e.destid",
+		goqu.L("ARRAY[r.uid]").As("path"), goqu.L("1").As("level"))
+
+	selectNext := make([]interface{}, 0)
+	selectNext = append(selectNext, "r.uid", "r.data", "e.destkind", "e.sourceid", "e.destid",
+		goqu.L("sg.path||r.uid").As("path"), goqu.L("level+1").As("level"))
+
+	sql, params, err := goqu.From("search_graph").
+		WithRecursive("search_graph(uid, data, destkind, sourceid, destid, path, level)",
+			goqu.From(schema.Table("resources").As("r")).InnerJoin(schema.Table("edges").As("e"),
+				goqu.On(goqu.ExOr{"r.uid": []exp.IdentifierExpression{goqu.I("e.sourceid"), goqu.I("e.destid")}})).
+				Select(selectBase...).
+				Where(goqu.I("r.uid").In(s.uids)).
+				Union(goqu.From(schema.Table("resources").As("r")).InnerJoin(schema.Table("edges").As("e"),
+					goqu.On(goqu.Ex{"r.uid": goqu.I("e.sourceid")})).
+					InnerJoin(goqu.T("search_graph").As("sg"),
+						goqu.On(goqu.ExOr{"sg.destid": goqu.I("e.sourceid"), "sg.sourceid": goqu.I("e.destid")})).
+					Select(selectNext...).
+					Where(goqu.Ex{"sg.level": goqu.L("1"),
+						"r.uid": goqu.Op{"neq": goqu.All("{sg.path}")}}))).
+		Select("data", "destid", "destkind").Distinct("destid").
+		Where(whereDs...).ToSQL()
+
+	if err != nil {
+		klog.Error("Error creating relation query", err)
```

Comment:
I wonder if we are able to continue after this error.

Reply:
Would the client receive an error? GraphQL can send partial results and errors.
Also the error we send to the client shouldn't describe any internals like the SQL query.

**We can move this discussion out of this PR, we should review the error path for the entire project.**

Reply:
Yes, we will revisit the error path for the entire api along with the context.